tests/lib/rules/no-redeclare.js

```diff
@@ -716,3 +716,640 @@ ruleTester.run("no-redeclare", rule, {
+		{
+			code: "/*global b:false*/ var b = 1;",
+			options: [{ builtinGlobals: true }],
+			errors: [
+				{
+					data: {
+						id: "b",
+					},
+					messageId: "redeclaredBySyntax",
+					type: "Block",
+				},
+			],
+		},
+
+		{
+			code: `
+	  type T = 1;
+	  type T = 2;
+			`,
+			errors: [
+				{
+					data: {
+						id: "T",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  type NodeListOf = 1;
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: {
+				parserOptions: {
+					lib: ["dom"],
+					sourceType: "script",
+				},
+			},
+			errors: [
+				{
+					data: {
+						id: "NodeListOf",
+					},
+					messageId: "redeclaredAsBuiltin",
+				},
+			],
+		},
+		{
+			code: `
+	  function A() {}
+	  class A {}
+			`,
+
+			errors: [
+				{
+					data: {
+						id: "A",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  function A() {}
+	  class A {}
+	  namespace A {}
+			`,
+			errors: [
+				{
+					data: {
+						id: "A",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+				{
+					data: {
+						id: "A",
+					},
+					line: 4,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  type something = string;
+	  const something = 2;
+			`,
+			errors: [
+				{
+					data: {
+						id: "something",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	    var Foo;
+		namespace Foo {
+			export const a = 2;
+		}
+		namespace Foo {
+			export const b = 4;
+		}
+			`,
+			errors: [
+				{
+					data: {
+						id: "Foo",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+				{
+					data: {
+						id: "Foo",
+					},
+					line: 6,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+			namespace Foo {
+				export const a = 1;
+			}
+
+			namespace Bar {
+				export const a = 1;
+			}
+			`,
+			languageOptions: {
+				globals: {
+					Foo: "readonly",
+					Bar: "readonly",
+				},
+			},
+			errors: [
+				{
+					data: {
+						id: "Foo",
+					},
+					line: 2,
+					messageId: "redeclaredAsBuiltin",
+				},
+				{
+					data: {
+						id: "Bar",
+					},
+					line: 6,
+					messageId: "redeclaredAsBuiltin",
+				},
+			],
+		},
+		{
+			code: `
+			/* global Foo */
+			namespace Foo {
+				export const a = 1;
+			}
+
+			/* global Bar */
+			namespace Bar {
+				export const a = 1;
+			}
+			`,
+			errors: [
+				{
+					data: {
+						id: "Foo",
+					},
+					line: 2,
+					messageId: "redeclaredBySyntax",
+				},
+				{
+					data: {
+						id: "Bar",
+					},
+					line: 7,
+					messageId: "redeclaredBySyntax",
+				},
+			],
+		},
+		{
+			code: `
+			namespace Foo {
+				export const a = 1;
+			}
+
+			/* global Bar */
+			namespace Bar {
+				export const a = 1;
+			}
+			`,
+
+			languageOptions: {
+				globals: {
+					Foo: "readonly",
+				},
+			},
```

Comment:
As above, I'm not sure if the combination of a value and type with the same name should be forbidden.